src/cairoUtilFuncGen/memory/implicitCoversion.ts

```diff
@@ -0,0 +1,287 @@
+import assert from 'assert';
+import {
+  ArrayType,
+  DataLocation,
+  Expression,
+  FunctionCall,
+  generalizeType,
+  getNodeType,
+  IntType,
+  PointerType,
+  TypeNode,
+} from 'solc-typed-ast';
+import { AST } from '../../ast/ast';
+import { CairoType, TypeConversionContext } from '../../utils/cairoTypeSystem';
+import { createCairoFunctionStub, createCallToFunction } from '../../utils/functionGeneration';
+import { narrowBigIntSafe, typeNameFromTypeNode } from '../../utils/utils';
+import { uint256 } from '../../warplib/utils';
+import { CairoFunction, StringIndexedFuncGen } from '../base';
+import { MemoryReadGen } from './memoryRead';
+import { MemoryWriteGen } from './memoryWrite';
+
+/*
+  Class that converts arrays with smaller element types into bigger types
+  e. g. 
+    uint8[] -> uint256[]
+    uint8[3] -> uint256[]
+    uint8[3] -> uint256[3]
+    uint8[3] -> uint256[8]
+  Only int/uint type implicit conversions
+*/
+export class MemoryImplicitConversionGen extends StringIndexedFuncGen {
+  constructor(private memoryWrite: MemoryWriteGen, private memoryRead: MemoryReadGen, ast: AST) {
+    super(ast);
+  }
+
+  genIfNecesary(targetExpression: Expression, sourceExpression: Expression): [Expression, boolean] {
+    const targetType = getNodeType(targetExpression, this.ast.compilerVersion);
+    const sourceType = getNodeType(sourceExpression, this.ast.compilerVersion);
+
+    const targetBaseType = getBaseType(targetType);
+    const sourceBaseType = getBaseType(sourceType);
+
+    const targetBaseCairoType = CairoType.fromSol(
+      targetBaseType,
+      this.ast,
+      TypeConversionContext.MemoryAllocation,
```

Comment:
I'm unconvinced about this, and think it might be the cause of some of the issues you mentioned. Types inside a memory array are allocated with TypeConversionContext Ref, only the top level uses TypeConversionContext Memory. Checking whether a conversion is necessary should probably use the solidity types rather than the cairo types